### PR TITLE
Robust Training example

### PR DESCRIPTION
**Pull request by fabianp**

### CODE REVIEW

File: examples/robust_training.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+# Initialize solver and parameters.
+solver = OptaxSolver(opt=optax.adam(1e-3), fun=loss_fun)
+rng = jax.random.PRNGKey(0)
+init_params = CNN().init(rng, jnp.ones([1, 28, 28, 1]))["params"]
+l2_regul = 1e-4
+
+# Run training loop.
+
+# Equivalent to:
```
Comment (mblondel) [inline examples/robust_training.py:85]: Remove this comment (I think it got copy-pasted from another example).

File: examples/robust_training.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
```
Comment (mblondel) [inline examples/robust_training.py:1]: Add license header.

File: examples/robust_training.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+params, state = solver.init(init_params)
+for it in range(200):
+  data = next(train_ds)
+  images = data['image'].astype(jnp.float32) / 255
+  labels = data['label']
+
+  def fsg_attack(image, label, epsilon=0.1):
+    """Fast sign-gradient attack on the L-infinity ball with radius epsilon.
+    
+    """
```
Comment (mblondel) [inline examples/robust_training.py:95]: Move the triple quote to the same line.

File: examples/robust_training.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    # comppute gradient of the loss wrt to the image
+    grad = jax.grad(loss_fun, argnums=2)(params, l2_regul, image, label)
+    adv_image = image + epsilon * jnp.sign(grad)
+    # clip the image to ensure pixels are between 0 and 1
+    return jnp.clip(adv_image, 0, 1)
+
+  #images_adv = jax.vmap(fsg_attack, in_axes=(0, 0))(images, labels)
```
Comment (mblondel) [inline examples/robust_training.py:102]: Remove (unless it's useful to keep it?).

File: examples/robust_training.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
+"""Robust Training in JAXOpt.
+
+The following code trains a convolutional neural network (CNN) to be robust
+with respect to the fast sign gradient (FSG) method, which is an attack
+on the L-infinity ball.
+
+
```
Comment (mblondel) [inline examples/robust_training.py:8]: If that's possible to do in just a few lines, I would explain the gist of the idea here.

File: examples/robust_training.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+from jax import numpy as jnp
+
+from flax import linen as nn
+import optax
+
+from jaxopt import GradientDescent
+from jaxopt import loss
+from jaxopt import OptaxSolver
+from jaxopt import tree_util
+
```
Comment (mblondel) [inline examples/robust_training.py:27]: Add one more line here.

File: examples/robust_training.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+on the L-infinity ball.
+
+
+References:
+  Goodfellow, Ian J., Jonathon Shlens, and Christian Szegedy. "Explaining
+  and harnessing adversarial examples." https://arxiv.org/abs/1412.6572
+"""
+
+import tensorflow_datasets as tfds
+from matplotlib import pyplot as plt
```
Comment (mblondel) [inline examples/robust_training.py:15]: Not used.

**Review by mblondel — COMMENTED** [review]
Thanks a lot Fabian. Some minor comments.

**Comment by fabianp** [comment]
Thanks @mblondel for the review. I believe I have now addressed them

**Pull request merged**